ws-server/src/server.rs

```diff
@@ -166,38 +171,69 @@ async fn background_task(socket: tokio::net::TcpStream, methods: Arc<Methods>, i
 	let (mut sender, mut receiver) = server.into_builder().finish();
 	let (tx, mut rx) = mpsc::unbounded::<String>();
 
+	// Send results back to the client.
 	tokio::spawn(async move {
 		while let Some(response) = rx.next().await {
 			let _ = sender.send_binary_mut(response.into_bytes()).await;
 			let _ = sender.flush().await;
 		}
 	});
 
-	let mut data = Vec::new();
+	let mut data = Vec::with_capacity(100);
+
+	// Look up the "method" (i.e. function pointer) from the registered methods and run it passing in
+	// the params from the request. The result of the computation is sent back over the `tx` channel and
+	// the result(s) are collected into a `String` and sent back over the wire.
+	let execute = move |tx: RpcSender, req: JsonRpcRequest| {
+		if let Some(method) = methods.get(&*req.method) {
+			let params = RpcParams::new(req.params.map(|params| params.get()));
+			if let Err(err) = (method)(req.id, params, &tx, conn_id) {
+				log::error!("execution of method call '{}' failed: {:?}, request id={:?}", req.method, err, req.id);
+				send_error(req.id, &tx, JsonRpcErrorCode::ServerError(-1).into());
+			}
+		} else {
+			send_error(req.id, &tx, JsonRpcErrorCode::MethodNotFound.into());
+		}
+	};
 
 	loop {
 		data.clear();
 
 		receiver.receive_data(&mut data).await?;
 
-		match serde_json::from_slice::<JsonRpcRequest>(&data) {
-			Ok(req) => {
-				let params = RpcParams::new(req.params.map(|params| params.get()));
-
-				if let Some(method) = methods.get(&*req.method) {
-					(method)(req.id, params, &tx, id)?;
-				} else {
-					send_error(req.id, &tx, JsonRpcErrorCode::MethodNotFound.into());
+		// For reasons outlined [here](https://github.com/serde-rs/json/issues/497), `RawValue` can't be used with
+		// untagged enums at the moment. This means we can't use an `SingleOrBatch` untagged enum here and have to try
+		// each case individually: first the single request case, then the batch case and lastly the error. For the
+		// worst case – unparseable input – we make three calls to [`serde_json::from_slice`] which is pretty annoying.
+		// Our [issue](https://github.com/paritytech/jsonrpsee/issues/296).
+		if let Ok(req) = serde_json::from_slice::<JsonRpcRequest>(&data) {
+			execute(&tx, req);
+		} else if let Ok(batch) = serde_json::from_slice::<Vec<JsonRpcRequest>>(&data) {
+			if !batch.is_empty() {
+				// Batch responses must be sent back as a single message so we the results from each request in the
```

Comment:
```suggestion
				// Batch responses must be sent back as a single message so we read the results from each request in the
```

not sure if read or get here, I think read makes sense though